src/main/java/dev/sigstore/rekor/client/RekorVerifier.java

```diff
@@ -90,6 +92,66 @@ public void verifyEntry(RekorEntry entry)
    * @throws RekorVerificationException if the entry cannot be verified
    */
   public void verifyInclusionProof(RekorEntry entry) throws RekorVerificationException {
-    throw new UnsupportedOperationException("Verifying inclusion proof is not yet supported");
+
+    var inclusionProof =
+        entry
+            .getVerification()
+            .getInclusionProof()
+            .orElseThrow(
+                () ->
+                    new RekorVerificationException(
+                        "No inclusion proof was found in the rekor entry"));
+
+    var leafHash =
+        Hashing.sha256()
+            .hashBytes(combineBytes(new byte[] {0x00}, Base64.getDecoder().decode(entry.getBody())))
+            .asBytes();
+
+    // see: https://datatracker.ietf.org/doc/rfc9162/ section 2.1.3.2
+    var fn = inclusionProof.getLogIndex();
+    var sn = inclusionProof.getTreeSize() - 1;
+    var r = leafHash;
+    var hashes = inclusionProof.getHashes();
+
+    for (var hash : hashes) {
+      byte[] p = Hex.decode(hash);
+      if (sn == 0) {
+        throw new RekorVerificationException("Inclusion proof failed, ended prematurely");
+      }
+      if (fn == sn || fn % 2 == 1) {
+        r = hashChildren(p, r);
+        while (fn % 2 == 0) {
+          fn = fn >> 1;
+          sn = sn >> 1;
+        }
+      } else {
+        r = hashChildren(r, p);
+      }
+      fn = fn >> 1;
+      sn = sn >> 1;
+    }
+
+    var calcuatedRootHash = Hex.toHexString(r);
+    if (!calcuatedRootHash.equals(inclusionProof.rootHash())) {
+      throw new RekorVerificationException(
+          "Calculated inclusion proof root hash does not match provided root hash\n"
+              + calcuatedRootHash
+              + "\n"
+              + inclusionProof.rootHash());
+    }
```

Comment:
should we be verifying the logid using the reko public key?

Reply:
yeah probs. I think I have a TODO somewhere else about that: https://github.com/sigstore/sigstore-java/blob/main/src/main/java/dev/sigstore/rekor/client/RekorVerifier.java#L69